tests/helper.py

```diff
@@ -82,3 +93,130 @@ def dotproduct_adjointness_test(
+    ----------
+    linear_operator
+        linear operator
+    u
+        element of the domain of the operator
+    v
+        element of the range of the operator
+    relative_tolerance
+        default is pytorch's default for float16
+    absolute_tolerance
+        default is pytorch's default for float16
+
+    Raises
+    ------
+    AssertionError
+        if the gradient is not the adjoint
+
+
+    """
+    # Gradient of the forward via vjp
+    (_, vjpfunc) = torch.func.vjp(linear_operator.forward, u)
+    assert torch.allclose(
+        vjpfunc((v,))[0], linear_operator.adjoint(v)[0], rtol=relative_tolerance, atol=absolute_tolerance
+    )
+
+    # Gradient of the adjoint via vjp
+    (_, vjpfunc) = torch.func.vjp(linear_operator.adjoint, v)
+    assert torch.allclose(
+        vjpfunc((u,))[0], linear_operator.forward(u)[0], rtol=relative_tolerance, atol=absolute_tolerance
+    )
+
+
+def forward_mode_autodiff_of_linear_operator_test(
+    linear_operator: LinearOperator,
+    u: torch.Tensor,
+    v: torch.Tensor,
+    relative_tolerance: float = 1e-3,
+    absolute_tolerance=1e-5,
+):
+    """Test the forward-mode autodiff calculation.
+
+    Verifies that the Jacobian-vector product (jvp) is equivalent to applying the operator.
+
+    Note: This property should hold for all u and v.
+    Commonly, this function is called with two random vectors u and v.
+
+
+    Parameters
+    ----------
+    linear_operator
+        linear operator
+    u
+        element of the domain of the operator
+    v
+        element of the range of the operator
+    relative_tolerance
+        default is pytorch's default for float16
+    absolute_tolerance
+        default is pytorch's default for float16
+
+    Raises
+    ------
+    AssertionError
+        if the jvp yields different results than applying the operator
+
+
+    """
+    # jvp of the forward
+    assert torch.allclose(
+        torch.func.jvp(linear_operator.forward, (u,), (u,))[0][0],
+        linear_operator.forward(u)[0],
+        rtol=relative_tolerance,
+        atol=absolute_tolerance,
+    )
+
+    # jvp of the adjoint
+    assert torch.allclose(
+        torch.func.jvp(linear_operator.adjoint, (v,), (v,))[0][0],
+        linear_operator.adjoint(v)[0],
+        rtol=relative_tolerance,
+        atol=absolute_tolerance,
+    )
+
+
+def autodiff_of_operator_test(
+    operator: Operator[*Tin, tuple[torch.Tensor,]],
+    *u: Any,
+):
+    """Test autodiff an operator is working.
+
+    This test does not check that the gradient is correct but simply that it can be calculated using autodiff.
+
+    Parameters
+    ----------
+    operator
+        operator
+    u
+        element(s) of the domain of the operator
+
+    Raises
+    ------
+    AssertionError
+        if autodiff fails
+
+
+    """
+    # Forward-mode autodiff using jvp
+    assert torch.func.jvp(operator.forward, u, u)
+
+    # Backward-mode autodiff using vjp
+    assert torch.func.vjp(operator.forward, *u)
```

Comment:
torch.func.vjp(operator.forward, *u) does afaik not actually start the gradient calculation.
it returns function, that takes an element of the range and calculates the vjp.
you can use the frist output of  the torch.func.jvp(operator.forward, u, u) above , which will be an element of the range,